Framework/DataHandling/src/MaskDetectors.cpp

```diff
@@ -269,83 +399,143 @@ void MaskDetectors::execPeaks(PeaksWorkspace_sptr WS) {
  * @param spectraList :: A list of spectra numbers
  * @param WS :: The input workspace to be masked
+ * @param range_info :: tuple containing the range of spectra to process and
+ *                      Boolean indicating if these ranges are defined
  */
 void MaskDetectors::fillIndexListFromSpectra(
     std::vector<size_t> &indexList, const std::vector<specnum_t> &spectraList,
-    const API::MatrixWorkspace_sptr WS) {
-  // Convert the vector of properties into a set for easy searching
-  std::set<int64_t> spectraSet(spectraList.begin(), spectraList.end());
-  // Next line means that anything in Clear the index list first
-  indexList.clear();
-  indexList.reserve(WS->getNumberHistograms());
-
-  for (int i = 0; i < static_cast<int>(WS->getNumberHistograms()); ++i) {
-    const specnum_t currentSpec = WS->getSpectrum(i).getSpectrumNo();
-    if (spectraSet.find(currentSpec) != spectraSet.end()) {
-      indexList.push_back(i);
+    const API::MatrixWorkspace_sptr WS,
+    const std::tuple<size_t, size_t, bool> &range_info) {
+
+  std::vector<size_t> tmp_index;
+  size_t startIndex = std::get<0>(range_info);
+  size_t endIndex = std::get<1>(range_info);
+  bool range_constrained = std::get<2>(range_info);
+
+  if (range_constrained) {
+    constrainIndexInRange(indexList, tmp_index, startIndex, endIndex);
+  } else {
+    tmp_index.swap(indexList);
+  }
+
+  auto SpecID2IndMap = WS->getSpectrumToWorkspaceIndexMap();
+  for (auto it = spectraList.begin(); it != spectraList.end(); it++) {
+    auto element = SpecID2IndMap.find(*it);
+    if (element == SpecID2IndMap.end()) {
+      continue;
+    }
+    size_t ws_index = element->second;
+
+    if (range_constrained && (ws_index < startIndex || ws_index > endIndex)) {
+      continue;
     }
+    tmp_index.push_back(ws_index);
   }
+
+  tmp_index.swap(indexList);
 }
 
-/**
- * Append the indices of the masked spectra from the given workspace list to the
+/* Append the indices of the masked spectra from the given workspace list to the
  * given list
- * @param indexList :: An existing list of indices
- * @param maskedWorkspace :: An workspace with masked spectra
- */
+ *
+ * @param indexList :: An existing list of indices.
+ * @param sourceWS  :: An workspace with masked spectra.
+ * @param range_info :: tuple containing the range of spectra to process and
+ *                            Boolean indicating if these ranges are defined
+*/
 void MaskDetectors::appendToIndexListFromWS(
-    std::vector<size_t> &indexList,
-    const MatrixWorkspace_sptr maskedWorkspace) {
-  // Convert the vector of properties into a set for easy searching
-  std::set<int64_t> existingIndices(indexList.begin(), indexList.end());
-  int endIndex = getProperty("EndWorkspaceIndex");
-  if (endIndex == EMPTY_INT())
-    endIndex = static_cast<int>(maskedWorkspace->getNumberHistograms() - 1);
-  int startIndex = getProperty("StartWorkspaceIndex");
+    std::vector<size_t> &indexList, const MatrixWorkspace_sptr sourceWS,
```

Comment:
initial matrixWorkspace_sptr is not const -- I just transferring it as const to the function. Compiler recognises  MatrixWorkspace_const_sptr as different type. Would not work